### PR TITLE
Cache external calls

### PR DESCRIPTION
**Pull request by ptsayli**

### CODE REVIEW

File: helper/call.go [hunk 1]
```diff
@@ -302,24 +313,42 @@ func (c *ContractCaller) IsTxConfirmed(currentTime time.Time, tx common.Hash) bo
...
+	if ok == false {
+		// get main chain block
+		receiptBlock, err := c.GetMainChainBlock(receipt.BlockNumber)
+		if err != nil {
+			Logger.Error("error getting receipt block from main chain", "Error", err)
+			return nil, err
+		}
+		Logger.Debug("Receipt block on main chain obtained", "Block", receiptBlock.Number.Uint64())
 
+		c.BlockTimeCache.Add(receipt.BlockNumber.Uint64(), receiptBlock.Time)
```
Comment (mankenavenkatesh) [inline helper/call.go]: time is not updated when cache miss. 

`time := receiptBlock.Time`

File: helper/call.go [hunk 2]
```diff
@@ -302,24 +313,42 @@ func (c *ContractCaller) IsTxConfirmed(currentTime time.Time, tx common.Hash) bo
 
 // GetConfirmedTxReceipt returns confirmed tx receipt
 func (c *ContractCaller) GetConfirmedTxReceipt(currentTime time.Time, tx common.Hash) (*ethTypes.Receipt, error) {
-	// get main tx receipt
-	receipt, err := c.GetMainTxReceipt(tx)
-	if err != nil {
-		Logger.Error("Error while fetching mainchain receipt", "error", err, "txHash", tx.Hex())
-		return nil, err
+	err := error(nil)
```
Comment (mankenavenkatesh) [inline helper/call.go]: initialization is not required.

File: helper/call.go [hunk 3]
```diff
@@ -302,24 +313,42 @@ func (c *ContractCaller) IsTxConfirmed(currentTime time.Time, tx common.Hash) bo
...
+	ok := false
```
Comment (mankenavenkatesh) [inline helper/call.go]: initialization is not required.

File: helper/call.go [hunk 4]
```diff
@@ -302,24 +313,42 @@ func (c *ContractCaller) IsTxConfirmed(currentTime time.Time, tx common.Hash) bo
...
+	time := uint64(0)
+	receipt := &ethTypes.Receipt{}
+	receiptCache, ok := c.ReceiptCache.Get(tx.String())
+
+	if ok == false {
```
Comment (jdkanani) [inline helper/call.go:322]:
```suggestion
	if !ok {
```

File: helper/call.go [hunk 5]
```diff
@@ -101,6 +105,8 @@ func NewContractCaller() (contractCallerObj ContractCaller, err error) {
 	contractCallerObj.MainChainClient = GetMainClient()
 	contractCallerObj.MaticChainClient = GetMaticClient()
 	contractCallerObj.MainChainRPC = GetMainChainRPCClient()
+	contractCallerObj.ReceiptCache = NewLru(100)
```
Comment (jdkanani) [inline helper/call.go:108]:
```suggestion
	contractCallerObj.ReceiptCache = NewLru(1000)
```
Increase to 1000

File: helper/call.go [hunk 6]
```diff
@@ -101,6 +105,8 @@ func NewContractCaller() (contractCallerObj ContractCaller, err error) {
...
+	contractCallerObj.BlockTimeCache = NewLru(100)
```
Comment (jdkanani) [inline helper/call.go:109]:
```suggestion
	contractCallerObj.BlockTimeCache = NewLru(1000)
```
Increase to 1000

File: helper/call.go [hunk 7]
```diff
@@ -169,6 +175,11 @@ func NewContractCaller() (contractCallerObj ContractCaller, err error) {
 	return
 }
 
+func NewLru(size int) *lru.Cache {
+	lruObj, _ := lru.New(size)
```
Comment (jdkanani) [inline helper/call.go:179]: Handle error

File: helper/call.go [hunk 8]
```diff
@@ -302,24 +313,42 @@ func (c *ContractCaller) IsTxConfirmed(currentTime time.Time, tx common.Hash) bo
...
+	if ok == false {
+		// get main tx receipt
+		receipt, err = c.GetMainTxReceipt(tx)
+		if err != nil {
+			Logger.Error("Error while fetching mainchain receipt", "error", err, "txHash", tx.Hex())
+			return nil, err
+		}
+		c.ReceiptCache.Add(tx.String(), receipt)
+	} else {
+		receipt, ok = receiptCache.(*ethTypes.Receipt)
```
Comment (jdkanani) [inline helper/call.go:331]: Handle if `ok` is false (`receiptCache` is not receipt type)

File: helper/call.go [hunk 9]
```diff
@@ -353,25 +367,43 @@ func (c *ContractCaller) IsTxConfirmed(currentTime time.Time, tx common.Hash, tx
 
 // GetConfirmedTxReceipt returns confirmed tx receipt
 func (c *ContractCaller) GetConfirmedTxReceipt(currentTime time.Time, tx common.Hash, txConfirmationTime time.Duration) (*ethTypes.Receipt, error) {
-	// get main tx receipt
-	receipt, err := c.GetMainTxReceipt(tx)
-	if err != nil {
-		Logger.Error("Error while fetching mainchain receipt", "error", err, "txHash", tx.Hex())
-		return nil, err
+	time := uint64(0)
+	receipt := &ethTypes.Receipt{}
```
Comment (jdkanani) [inline helper/call.go:371]:
```suggestion
	var receipt *ethTypes.Receipt = nil
```

File: helper/call.go [hunk 10]
```diff
@@ -353,25 +367,43 @@ func (c *ContractCaller) IsTxConfirmed(currentTime time.Time, tx common.Hash, tx
...
+	receiptCache, ok := c.ReceiptCache.Get(tx.String())
+
+	if !ok {
+		// get main tx receipt
+		fetchedReceipt, err := c.GetMainTxReceipt(tx)
+		receipt = fetchedReceipt
```
Comment (jdkanani) [inline helper/call.go:377]:
```suggestion
                var err error
		receipt, err = c.GetMainTxReceipt(tx)
```

File: helper/call.go [hunk 11]
```diff
@@ -353,25 +367,43 @@ func (c *ContractCaller) IsTxConfirmed(currentTime time.Time, tx common.Hash, tx
...
+		if err != nil {
+			Logger.Error("Error while fetching mainchain receipt", "error", err, "txHash", tx.Hex())
+			return nil, err
+		}
+		c.ReceiptCache.Add(tx.String(), receipt)
+	} else {
+		receiptCache, _ := receiptCache.(*ethTypes.Receipt)
+		receipt = receiptCache
```
Comment (jdkanani) [inline helper/call.go:385]:
```suggestion
		receipt, _ = receiptCache.(*ethTypes.Receipt)
```

File: helper/call.go [hunk 12]
```diff
@@ -353,25 +367,45 @@ func (c *ContractCaller) IsTxConfirmed(currentTime time.Time, tx common.Hash, tx
...
+		var err error
+
+		// get main tx receipt
+		receipt, err = c.GetMainTxReceipt(tx)
+		if err != nil {
+			Logger.Error("Error while fetching mainchain receipt", "error", err, "txHash", tx.Hex())
+			return nil, err
+		}
+
+		c.ReceiptCache.Add(tx.String(), receipt)
```
Comment (mankenavenkatesh) [inline helper/call.go:385]: won't it be sufficient to store only receipt.BlockNumber.Uint64() in cache?

Comment (ptsayli) [reply helper/call.go:385]: are we sure that, we won't require other field from Receipt in future?
then we can only store block number
That will be better

Comment (jdkanani) [reply helper/call.go:385]: We need receipt.

**Pull request merged**